### DISCUSSION TITLE
New Notifications support ?

### DISCUSSION THREAD

**Discussion by luc-github** [luc-github/ESP3D#549] [Ideas]
Is there any service that would be interresting ?
* Discord 
using webHook it seems not difficult to implement 
* Alexa from Amazon
Need to do some research if can be implemented and what need to be done to be implemented
https://www.youtube.com/watch?v=WKn5ynGXdoI
https://arduino.blaisepascal.fr/mqtt-avec-arduino/
http://www.steves-internet-guide.com/using-arduino-pubsub-mqtt-client/
ho boy.... https://developer.amazon.com/en-US/docs/alexa/sdns/use-skill-development-notifications.html
* Notify to WebUI, send popup  with message to webUI if open
* TBD

**Comment by AndyRockall** [comment]
MQTT would be nice

**Reply by luc-github** [reply]
this need to setup a MQTT broker also no ?

**Reply by luc-github** [reply]
@AndyRockall do you have an user case?
MQTT is a more a communication protocol than notification itself per my understanding of https://randomnerdtutorials.com/what-is-mqtt-and-how-it-works/

**Reply by realroywalker** [reply]
+1 for MQTT no broker needed locally (on esp) only client library. It would be nice to have as once the mqtt message is on the queue it's possible to forward out to many other notification types (as the user decides)
One example - esp3d just fires mqtt message to say print done, homeassistant (as an example) consumes the message and triggers a smart switch to power off the printer, it also triggers an audio notification (via Alexa or Google home) to say the print finished.
This way you don't need any special Alexa/Google skill, since the user picks their own method based on what they need.

**Reply by luc-github** [reply]
@realroywalker , thank you - I am not familiar with MQTT
if I refer to https://www.valvers.com/open-software/arduino/esp32-mqtt-tutorial/
what you suggest is : 
in settings put broker address and port: T1=broker.shiftr.io T2=8883
so we can setup the `client client.begin("broker.shiftr.io", 8883, net);` properly
then 
using [ESP610] Ts="/hello" parameter would be the path 
so in [ESP600] world , we just need to provide message for `client.publish("/hello", "world");` 
if need a new path just use [ESP610]TS=/newpath before using [ESP600]

Is that what your suggest ? - I think it could be done easily

**Reply by realroywalker** [reply]
Hi @luc-github ,
Yes pretty much, the mqtt could either be local network or some cloud service so having a few setting for the user to choose would be all that's needed. Then some small code for publishing to a topic when a notification happens.
This site has a good guide https://techtutorialsx.com/2017/04/09/esp8266-connecting-to-mqtt-broker
I think the minimum settings needed are: 
- Broker address /port
- username & password (if Auth enabled)
- client Id (can just be made up by the user)
- topic name

**Reply by luc-github** [reply]
what is `topic name` ? is that the path of published notification ?

**Reply by realroywalker** [reply]
This is the topic on the broker that the messages would be posted to. So that any service/application can subscribe to a topic they are want to consume.
Think of the broker as one of those apartment block mailboxes with slots for each apartment. In that case each individual apartment would be a "topic" (”esp3d” for example) then for any programs that I'd like to make use of esp3d notifications, I tell the program to subscribe to the "esp3d" topic. Now when some "mail" (eg. Notifications sent by esp3d) arrives to that slot, the program would see the notification and action it. (Like taking the mail from the slot and reading it in the mail box analogy)

**Reply by luc-github** [reply]
Notification has 3 settings:
T1 limited to 64Bytes
T2 limited to 64Bytes
TS: limited to 128Bytes
Notification Message itself

I would like to avoid to add more setting just for MQTT 

* so Authentication can be done using T1/T2
* broker address/port using TS with a `:` separator  between server address and port
* client id would be the hostname, so no need a new setting
* topic and message would be in [ESP600] command like `[ESP600]<topic>=<message>` 

One issue I see, is the auto notification feature which currently send the IP of ESP3D at start but also send error  and more in future
I am not sure how to handle it, force predefined topic (notification=ip:XXX.XXX.XXX.XXX) ? or just ignore it ?

**Comment by luc-github** [comment]
Added Alexa per @AndyRockall suggestion

**Comment by luc-github** [comment]
add webUI popup - this one should work in parallele of others

**Comment by KingBain** [comment]
I think it would be cool to support API based notification services
https://documentation.notification.canada.ca/en/
https://docs.notifications.service.gov.uk/rest-api.html#rest-api-documentation

**Comment by luc-github** [comment]
these services seems not free, are they ?

**Reply by luc-github** [reply]
One request from Discord is to send to WebUI, may be using the new toaster method currently used on for error ?

![image](https://user-images.githubusercontent.com/8822552/152078222-f4452a39-4503-4082-8940-94a1f44cf7b6.png)

**Reply by KingBain** [reply]
> One request from Discord is to send to WebUI, may be using the new toaster method currently used on for error ?

IDK about discord, but if its similar to Slack and Teams; they both support "incoming webhooks".

You could use the IFTTT connector to push messages to a incoming webhooks url.

**Reply by luc-github** [reply]
yes I use webhook with discord in CI process to raise notification to ESP3D discord server
The suggestion of WebUi notification was raised in ESP3D discord server, I have added screenshot for reference.
